server/parsec/components/postgresql/invite.py

```diff
@@ -1405,95 +1410,72 @@ async def get_active_attempt(
     ) -> GreetingAttemptID:
-        (
-            greeting_attempt_internal_id,
-            greeting_attempt_id,
-        ) = await self.get_active_attempt(
+        # Get the greeting session
+        greeting_session_id = await self.get_greeting_session(
             conn,
             organization_id=organization_id,
             greeter=greeter,
             invitation_internal_id=invitation_internal_id,
         )
-        is_active = await self.greeter_join_or_cancel(conn, greeting_attempt_internal_id, now)
-        while not is_active:
-            (
-                greeting_attempt_internal_id,
-                greeting_attempt_id,
-            ) = await self.get_active_attempt(
-                conn,
-                organization_id=organization_id,
-                greeter=greeter,
-                invitation_internal_id=invitation_internal_id,
-            )
-            is_active = await self.greeter_join_or_cancel(conn, greeting_attempt_internal_id, now)
-        return greeting_attempt_id
-
-    async def new_attempt_for_claimer(
-        self,
-        conn: AsyncpgConnection,
-        organization_id: OrganizationID,
-        greeter: UserID,
-        invitation_internal_id: int,
-        now: DateTime,
-    ) -> GreetingAttemptID:
+        # Get the active attempt
         (
             greeting_attempt_internal_id,
             greeting_attempt_id,
-        ) = await self.get_active_attempt(
+        ) = await self.get_active_greeting_attempt(
             conn,
             organization_id=organization_id,
-            greeter=greeter,
-            invitation_internal_id=invitation_internal_id,
+            greeting_session_id=greeting_session_id,
         )
-        is_active = await self.claimer_join_or_cancel(conn, greeting_attempt_internal_id, now)
-        while not is_active:
+        # Try to join the attempt
+        is_active = await self.join_or_cancel(peer, conn, greeting_attempt_internal_id, now)
+        # The attempt has been joined
+        if is_active:
+            return greeting_attempt_id
+        # The attempt was already joined and has been cancelled
+        else:
+            # Get a fresh attempt
             (
                 greeting_attempt_internal_id,
                 greeting_attempt_id,
-            ) = await self.get_active_attempt(
+            ) = await self.get_active_greeting_attempt(
                 conn,
                 organization_id=organization_id,
-                greeter=greeter,
-                invitation_internal_id=invitation_internal_id,
+                greeting_session_id=greeting_session_id,
             )
-            is_active = await self.claimer_join_or_cancel(conn, greeting_attempt_internal_id, now)
-        return greeting_attempt_id
+            # Join the attempt
+            is_active = await self.join_or_cancel(peer, conn, greeting_attempt_internal_id, now)
+            # It has to be active since we just created it
+            assert is_active
```

Comment:
From what I understand, asserts are not checked if code is optimized, and as attempt creation and joining are not atomic, should we add a warning or a real error ?

Reply:
We never disable the asserts when running the server, so asserts are used for errors that would indicate a bug. They are logged by sentry and properly managed by the backend (hopefully)